fennel/test_lib/data_engine.py

```diff
@@ -0,0 +1,645 @@
+            empty_df = pd.DataFrame(
+                columns=val_cols, data=[[None] * len(val_cols)] * len(keys)
+            )
+            return empty_df, pd.Series(np.array([False] * len(keys)))
+
+        timestamp_field = self.datasets[cls_name].dataset.timestamp_field
+        timestamp_length = len(ts)
+        if timestamp_length != keys.shape[0]:
+            raise ValueError(
+                f"Length of timestamp array `{timestamp_length}` does not match ",
+                f"length of keys array {keys.shape[0]} for dataset {cls_name}.",
+            )
+        keys = keys.reset_index(drop=True)
+        ts = ts.reset_index(drop=True)
+        assert keys.shape[0] == len(
+            ts
+        ), "Length of keys and ts should be same " "found {} and {}".format(
+            keys.shape[0], len(ts)
+        )
+        keys[timestamp_field] = ts
+
+        keys[FENNEL_ORDER] = np.arange(len(keys))
+        # Sort the keys by timestamp
+        keys = keys.sort_values(timestamp_field)
+        if self.datasets[cls_name].aggregated_datasets:
+            data_dict = self.datasets[cls_name].aggregated_datasets
+            # Gather all the columns that are needed from data_dict to create a df.
+            result_dfs = []
+            for col, right_df in data_dict.items():  # type: ignore
+                try:
+                    df = self._dataframe_lookup(
+                        cls_name, keys, right_df, join_columns, timestamp_field
+                    )
+                except ValueError as err:
+                    raise ValueError(err)
+                df = df.set_index(FENNEL_ORDER).loc[np.arange(len(df)), :]
+                result_dfs.append(df)
+            # Get common columns
+            common_columns = set(result_dfs[0].columns)
+            for df in result_dfs[1:]:
+                common_columns.intersection_update(df.columns)
+
+            # Remove common columns from all DataFrames except the first one
+            for i in range(1, len(result_dfs)):
+                result_dfs[i] = result_dfs[i].drop(columns=common_columns)
+
+            # Concatenate the DataFrames column-wise
+            df = pd.concat(result_dfs, axis=1)
+        else:
+            right_df = self.datasets[cls_name].data
+            try:
+                df = self._dataframe_lookup(
+                    cls_name, keys, right_df, join_columns, timestamp_field
+                )
+            except ValueError as err:
+                raise ValueError(err)
+            df.rename(columns={FENNEL_TIMESTAMP: timestamp_field}, inplace=True)
+            df = df.set_index(FENNEL_ORDER).loc[np.arange(len(df)), :]
+
+        found = df[FENNEL_LOOKUP].apply(lambda x: x is not np.nan)
+        df.drop(columns=[FENNEL_LOOKUP], inplace=True)
+        if len(fields) > 0:
+            df = df[fields]
+        df = df.reset_index(drop=True)
+        return df, found
+
+    def _dataframe_lookup(
```

Comment:
dataframe_lookup sounds wierd? call it as_of_lookup?

Reply:
changed